File: 7/7.rb
# you can use here mathn lib, but I won't

@primes = [2]
n = 0

while n != 10000
  i = @primes.last
  is_prime = false

  while !is_prime
    i += 1
    @primes.each do |prime|
      is_prime = i % prime != 0
      break unless is_prime
    end
  end

  @primes << i
  n += 1
end
p @primes.last

File: 7/7.ml
open Core.Std

let is_prime number primes =
  List.for_all primes ~f:(fun x -> (number mod x) <> 0)

let rec next_prime primes = function
  | current when is_prime current primes -> current
  | current -> next_prime primes (current+2)

let find_prime n =
  let rec find_prime_rec primes last count = match count with
    | 0 -> last
    | _ -> let prime = next_prime primes (last+2) in
             find_prime_rec (primes@[prime]) prime (count-1)
  in
    find_prime_rec [2; 3] 3 (n-2)

let () =
  Printf.printf "%d\n" (find_prime 10001)
